neps/nep-504.md

```diff
@@ -0,0 +1,242 @@
+
+## Specification
+
+NEP-393 `TokenMetadata` is minimalistic. It only requires data which a contract can reason about, and all user facing details (such as title, description) are meant to be part of the reference.
+Similarly, `ContractMetadata` only defines issuer basic info.
+
+### Class Metadata
+
+We start by defining a `ClassMetadata` which will contain all common information about tokens of the same class.
+This will allow to avoid duplicating such information in all `TokenMetadata` and potentially allow to leave `TokenMetadata.reference` empty, if the tokens only differentiate by the class, issued_at and expires_at attributes.
+
+```rust
+pub struct ClassMetadata {
+    /// Issuer class name. Required.
+    pub name: String,
+    /// If defined, should be used instead of `contract_metadata.symbol`.
+    pub symbol: Option<String>,
+    /// Icon content (SVG) or a link to an Icon. If it doesn't start with a scheme (eg: https://)
+    /// then `contract_metadata.base_uri` should be prepended.
+    pub icon: Option<String>,
+    /// JSON or an URL to a JSON file with more info. If it doesn't start with a scheme
+    /// (eg: https://) then base_uri should be prepended.
+    pub reference: Option<String>,
+    /// Base64-encoded sha256 hash of JSON from reference field. Required if `reference` is included.
+    pub reference_hash: Option<Base64VecU8>,
+}
+```
+
+The Class Metadata should be defined by the Issuer and stored by the issuer contract. Note, that Contract Metadata is stored in the issuer contract and the Token Metadata is stored in the registry.
+Having Class Metadata in issuer contract allows more granular storage control, shard the data, and facilitate easier migration patterns.
+
+We recommend to store in the Class Metadata as much common information as possible, rather than pushing it down to the Token Metadata. For example if all tokens of give (issuer, class) pair have the same `icon`, then the `icon` should be defined in the `ClassMetadata` rather than `TokenMetadata.reference`.
+
+The API to set `ClassMetadata` is not part of the standard, and can be custom for every implementation.
+
+### Reference Field
+
+The `reference` is attribute is part of all Metadata structures (Token, Class and Contract). It must be a valid JSON string or an URL link to a valid JSON file. In the latter case we recommend to set `reference_hash` to facilitate secure off-chain metadata storage and a mechanism for consistency checks.
+It's up to the issuer if he wants to store `reference` fully on chain (as a JSON string) or off-chain. The latter case won't guarantee data availability, but may provide faster data access.
+
+We define a `ReferenceBase` type. All entries are optional, and don't have to be set. If an entry is set, then it must have type compatible with `ReferenceBase`. The Reference object can extend the `ReferenceBase` and provide more fields.
+
+```rust
+struct ReferenceBase {
```

Comment:
The second big question is: should the reference field be standardized at all? The structure fields look like taken from a concrete/specialized project, and all of the fields are... optional.

Reply:
I also don't understand why these fields that are proposed to be standardized are part of `reference`, which was designed to be a link to off-chain additional information.

> It must be a valid JSON string or ...

This is an abuse of the `reference` field. Just add more fields where you need them. That is, I suggest putting these fields on the `*Metadata` level as individual fields instead of creating unnecessary nesting. Metadata structs can have any number of additional fields on them, and that does not contradict the minimum metadata structure as at the API level it is communicated through JSONs and additional fields can be easily ignored.